turbodbc dependency does not work for poetry based projects

## Description
Installation of `turbodbc` as dependency in a [poetry](https://python-poetry.org/) based project does not work. 

## Environment
```
OS: Ubuntu 21.10
Python: 3.9.7
Poetry: 1.1.13
```

## Steps to reproduce
1. [install poetry](https://python-poetry.org/docs/#installation)
2. create a new project
```shell
user@host ~ $ poetry new turbodbc_issue
```
3. `cd` into the newly created project
```shell
user@host ~ $ cd turbodbc_issue
```
4. add turbodbc as a dependency to the project
```shell
user@host ~/turbodbc_issue $ poetry add -vvv turbodbc
```

## Root cause
Missing/Incorrect linker flags for linkage of `turbodbc_intern.<platform>.so`


```
x86_64-linux-gnu-g++ -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-x86_64-cpython-39/src/turbodbc_python/determine_parameter_type.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/buffer_size.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/column_info.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/connect.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/connection.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/cursor.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/error.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/module.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/options.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/python_parameter_set.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_bindings/python_result_set.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_parameter_set.o build/temp.linux-x86_64-cpython-39/src/turbodbc_python/python_result_set.o -Lbuild/lib.linux-x86_64-3.9 -L/usr/lib/x86_64-linux-gnu -lodbc -lturbodbc.cpython-39-x86_64-linux-gnu -o build/lib.linux-x86_64-cpython-39/turbodbc_intern.cpython-39-x86_64-linux-gnu.so -Wl,-rpath,$ORIGIN
        /usr/bin/ld: cannot find -lturbodbc.cpython-39-x86_64-linux-gnu
        collect2: error: ld returned 1 exit status
        error: command '/usr/bin/x86_64-linux-gnu-g++' failed with exit code 1
        [end of output]

    note: This error originates from a subprocess, and is likely not a problem with pip.
  error: legacy-install-failure

  × Encountered error while trying to install package.
  ╰─> turbodbc

  note: This is an issue with the package mentioned above, not pip.
  hint: See above for output from the failure.
  WARNING: There was an error checking the latest version of pip.
```

`-L` points to `-Lbuild/lib.linux-x86_64-3.9` while the (temporary) build folder looks like this:
```
build
├── lib.linux-x86_64-cpython-39
│   ├── libturbodbc.cpython-39-x86_64-linux-gnu.so
│   └── turbodbc
│       ├── api_constants.py
│       ├── connection.py
...
```
It looks like the `linking path` is deduced by a [convention](https://github.com/blue-yonder/turbodbc/blob/c7b2389a1eaccb5ce6847d96a84d99cf5c42c2e5/setup.py#L22) before the build is started.
:warning: This may lead to similar build issues with other package mangers :warning: 


## Leads (Additional Information)
### setup.py
`setup.py  build` supports various options regarding the build directory so it may not be stable to assume specific paths.
```
user@host ~ $ python setup.py build -h
...
Options for 'build' command:
  --build-base (-b)  base directory for build library
  --build-purelib    build directory for platform-neutral distributions
  --build-platlib    build directory for platform-specific distributions
  --build-lib        build directory for all distribution (defaults to either
                     build-purelib or build-platlib
  --build-scripts    build directory for scripts
  --build-temp (-t)  temporary build directory
  --plat-name (-p)   platform name to build for, if supported (default: linux-
                     x86_64)
...
```
### Poetry
As far as I could see, poetry internally also uses `pip`, but disables its dependency resolution etc.

## Possible Solution(s)
* Provide wheels for common platforms (Mitigation)
* Find a way to reliably retrieve the build directory during the build and inject it appropriately
  (E.g. by adding custom builder/command)

## Related Issues
* https://github.com/exasol/sqlalchemy-exasol/issues/146

--- a/setup.py
+++ b/setup.py
@@ -7,8 +7,8 @@ from glob import iglob
 from typing import List
 
 import setuptools.command.build_ext
-from setuptools.command.build_ext import build_ext
 from setuptools import Distribution, Extension, setup
+from setuptools.command.build_ext import build_ext
 
 
 class TurbodbcExtensionBuilder(build_ext):
@@ -16,7 +16,7 @@ class TurbodbcExtensionBuilder(build_ext):
         super().run()
 
     def build_extension(self, extension: setuptools.extension.Extension) -> None:
-        extension.library_dirs.append(self.build_lib)
+        extension.library_dirs.append(self.build_lib)  # type: ignore
         super().build_extension(extension)
 
 

--- a/setup.py
+++ b/setup.py
@@ -3,12 +3,21 @@ import itertools
 import os
 import os.path
 import sys
-import sysconfig
 from glob import iglob
 from typing import List
 
 import setuptools.command.build_ext
+from setuptools.command.build_ext import build_ext
 from setuptools import Distribution, Extension, setup
+
+
+class TurbodbcExtensionBuilder(build_ext):
+    def run(self) -> None:
+        super().run()
+
+    def build_extension(self, extension: setuptools.extension.Extension) -> None:
+        extension.library_dirs.append(self.build_lib)
+        super().build_extension(extension)
 
 
 def _get_turbodbc_libname():
@@ -17,23 +26,6 @@ def _get_turbodbc_libname():
     without_lib = full_name.split("lib", 1)[-1]
     without_so = without_lib.rsplit(".so", 1)[0]
     return without_so
-
-
-def _get_distutils_build_directory():
-    """
-    Returns the directory distutils uses to build its files.
-    We need this directory since we build extensions which have to link
-    other ones.
-    """
-    pattern = "lib.{platform}-{major}.{minor}"
-    return os.path.join(
-        "build",
-        pattern.format(
-            platform=sysconfig.get_platform(),
-            major=sys.version_info[0],
-            minor=sys.version_info[1],
-        ),
-    )
 
 
 def _get_source_files(directory):
@@ -85,7 +77,7 @@ extra_compile_args = []
 hidden_visibility_args = []
 include_dirs = ["include/", _deferred_pybind11_include()]
 
-library_dirs = [_get_distutils_build_directory()]
+library_dirs = []
 python_module_link_args = []
 base_library_link_args: List[str] = []
 
@@ -225,7 +217,6 @@ here = os.path.abspath(os.path.dirname(__file__))
 with open(os.path.join(here, "README.md")) as f:
     long_description = f.read()
 
-
 setup(
     name="turbodbc",
     version="4.5.4",
@@ -259,5 +250,6 @@ setup(
         "Programming Language :: Python :: Implementation :: CPython",
         "Topic :: Database",
     ],
+    cmdclass=dict(build_ext=TurbodbcExtensionBuilder),
     ext_modules=get_extension_modules(),
 )

